README
------------------------------------------------------------------------------

This is the README file of the IgH EtherCAT Master.

$Id$

vim: spelllang=en spell tw=78

------------------------------------------------------------------------------

Contents:
1) General Information
2) Requirements
3) Building and installing
4) Realtime & Tuning
5) License
6) Coding Style
7) Synapticon-specific changes

------------------------------------------------------------------------------

1) General Information
======================

This is an open-source EtherCAT master implementation for Linux 2.6.

See the FEATURES file for a list of features. For more information, see
http://etherlab.org/en/ethercat.

or contact

Dipl.-Ing. (FH) Florian Pose <fp@igh-essen.com>
Ingenieurgemeinschaft IgH
Heinz-Baecker-Str. 34
D-45356 Essen
http://igh-essen.com

------------------------------------------------------------------------------

2) Requirements
===============

1) Software requirements

Configured sources for the Linux 2.6 kernel are required to build the EtherCAT
master.

2) Hardware requirements

A table of supported hardware can be found at
http://etherlab.org/en/ethercat/hardware.php.

------------------------------------------------------------------------------

3) Building and installing
==========================

See the INSTALL file.

------------------------------------------------------------------------------

4) Realtime & Tuning
====================

Realtime patches for the Linux kernel are supported, but not required. The
realtime processing has to be done by the calling module (see API
documentation). The EtherCAT master code itself is passive (except for the
idle mode and EoE).

To avoid frame timeouts, deactivating DMA access for hard drives is
recommended (hdparm -d0 <DEV>).

------------------------------------------------------------------------------

5) License
==========

Copyright (C) 2006-2009  Florian Pose, Ingenieurgemeinschaft IgH

This file is part of the IgH EtherCAT Master.

The IgH EtherCAT Master is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License version 2, as
published by the Free Software Foundation.

The IgH EtherCAT Master is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
details.

You should have received a copy of the GNU General Public License along with
the IgH EtherCAT Master; if not, write to the Free Software Foundation, Inc.,
51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

---

The license mentioned above concerns the source code only. Using the EtherCAT
technology and brand is only permitted in compliance with the industrial
property and similar rights of Beckhoff Automation GmbH.

------------------------------------------------------------------------------

6) Coding Style
===============

Developers shall use the coding style rules in the CodingStyle.txt file.

------------------------------------------------------------------------------

7) Synapticon-specific changes
==============================

This version of the IgH EtherCAT master contains changes made by Synapticon in
order to support its products.

All of the changes made from the original IgH EtherCAT master will be listed in
the CHANGELOG.md file.

Apart from the changes made to the base of this product, Synapticon has also
added some additional instructions, as well as installation and helper scripts.

------------------------------------------------------------------------------


NEWS
-------------------------------------------------------------------------------

$Id$

vim: spelllang=en spell tw=78

-------------------------------------------------------------------------------

Changes in 1.5.2:

* API extensions (find the complete description in include/ecrt.h)
    * Added redundancy features; enable using --with-devices.
    * Re-designed and seamlessly integrated RTDM interface.
    * Added ecrt_sdo_request_index() method to change index and subindex.
    * Changed the data types of the shift times in ecrt_slave_config_dc()
      to int32_t to correctly display negative shift times.
    * Added API for querying CoE emergency requests.
    * Added interface to read/write register contents; re-worked register
      requests.
    * Added interface to select the reference clock and to sync to it.
    * Exported ecrt_domain_size() to userspace.
    * Added ecrt_slave_config_reg_so_entry_pos() to register non-unique
      PDO entries.
* Ethernet drivers
    * Added 8139too driver for 3.0, 3.2, 3.4.
    * Added r8169 driver for 2.6.36 (J. Kunz), 3.2 (J. Kunz), 3.4 (F. Pose).
    * Added e1000 driver for 3.0, 3.4.
    * Added e1000e driver for 3.2 (J. Kunz), 3.4 (F. Pose).
    * Added e100 driver for 3.0, 3.4.
    * Fixes regarding 8139too driver for 2.6.36.
    * Some fixes for all e1000 and e1000e drivers.
* General behaviour
    * Avoided semaphore locking from userspace/rtdm library: Implemented
      datagram ring instead of queue.
    * Assign SII to PDI during PREOP transition; Introduced
      --enable-sii-assign to switch this on.
    * Avoided re-allocating memory in userspace library implementations of
      ecrt_sdo_request_state() and in VoE handlers.
    * Send sync datagrams only if reference clock is found.
    * Adjusted some DC clock discipline parameters.
    * Fixed AL status code lookup.
    * Fixed some ioctl() return values and permissions.
    * Fixed FoE data OpCode, thanks to R. Roesch.
    * Fixed and improved cancelling requests when clearing configuration.
    * Using common wait queue for requests.
    * Removed some unnecessary request types.
    * Output slave FSM datagram statistics.
    * Removed some state-dependent datagram skip messages.
    * Immediately restart slave FSM on several failed requests.
    * Reset master state machine when going to orphaned phase.
* Command-line tool:
    * Show complete access flag in config view.
    * Added EtherLab skin for 'ethercat pdos' command.
    * Fixed 'String too large' bug.
* Infrastructure:
    * Added AM_PROG_AR necessary for autoconf 1.12.
    * Avoided use of AM_CFLAGS, which is not used in some situations.
    * Added vmalloc() include fix for ARM architecture, thanks to Andrea
      Scian.
    * Install ethercat service to multi-user target by default.
    * Fixed COPYING.LESSER to be LGPL 2.1 (like in file headers).
    * More detailed output of configure script.
    * Added rtai_rtdm_dc example, thanks to Graeme Foot.

Changes in 1.5.1:

* Fixed reset of allow_scanning flag if ecrt_master_activate() was not called.
* Fixed missing distribution for r8169 for 2.6.32 and e1000 driver for 2.6.31.
* Added e100 for 2.6.35.
* Added r8169 driver for 2.6.35 (F. Pose).
* Added e1000e driver for 2.6.35 (F. Pose).
* Added fix for ESC port order (DC delay calculation).
* Added e1000 driver for 2.6.35.

Changes in 1.5.0:

* Added a userspace library for accessing the application interface. This
  library is licensed under LGPLv2.
* Added distributed clocks support.
* Added watchdog configuration via application interface (thanks to J. Mohre).
* Added VoE mailbox protocol support.
* Added SoE mailbox protocol support.
* Improved the callback mechanism. ecrt_master_callbacks() now takes two
  callback functions for sending and receiving datagrams.
  ecrt_master_send_ext() is used to execute the sending of non-application
  datagrams.
* Separated datagram initialization from filling the payload with zeros.
  Introduced new method ec_datagram_zero() for that.
* Added phy_read and phy_write commands to ethercat tool.
* Added e100 driver for Intel PRO/100 NICs.
    - Added e100 driver for 2.6.27.
    - Added e100 driver for 2.6.28, thanks to Kim. H. Madsen.
    - Added e100 driver for 2.6.29, thanks to Andre Puschmann.
    - Added e100 driver for 2.6.31.
    - Added e100 driver for 2.6.32.
    - Added e100 driver for 2.6.33, thanks to J. Kunz.
    - Added e100 driver for 2.6.37.
* Added 8139too driver for kernels 2.6.25 (F. Pose), 2.6.26 (M. Luescher),
  2.6.27, 2.6.28, 2.6.29 (M. Goetze), 2.6.31 (F. Pose), 2.6.32 (F. Pose),
  2.6.33 (J. Kunz), 2.6.34 (Malcolm Lewis), 2.6.35 (B. Benner),
  2.6.36 (F. Pose) and 2.6.37 (F. Pose).
* Added e1000 driver for 2.6.26 (M. Luescher), 2.6.27, 2.6.28, 2.6.29, 2.6.32,
  2.6.33 (J. Kunz), 2.6.35, 2.6.37.
* Added r8169 driver for 2.6.24, 2.6.27, 2.6.28, 2.6.29, 2.6.31, 2.6.32
  (Robbie K), 2.6.33 (J. Kunz), 2.6.35, 2.6.37.
* Added e1000e driver for 2.6.32, 2.6.33, 2.6.34 (thanks to Siwei Zhuang).
* Debug interfaces are created with the Ethernet addresses of the attached
  physical device.
* Improved error case return codes of many functions.
* Implemented the File Access over EtherCAT (FoE) mailbox protocol, thanks to
  O. Zarges.
* Going to the Bootstrap state is now supported by the state machines and the
  command-line tool.
* Added ecrt_open_master() and ecrt_master_reserve() separation for
  the userspace library (thanks to Martin Troxler).
* Added bus information interface (methods ecrt_master(),
  ecrt_master_get_slave(), ecrt_master_get_sync_manager(),
  ecrt_master_get_pdo() and ecrt_master_get_pdo_entry()) to get information
  about the currently connected slaves and the PDO entries provided (thanks to
  Martin Troxler and Stefan Weiser).
* Added ecrt_master_sdo_download() and ecrt_master_sdo_upload()
  methods to let an application transfer SDOs before activating the master
  (thanks to Stefan Weiser).

tool/main.cpp
/*****************************************************************************
 *
 *  $Id$
 *
 *  Copyright (C) 2006-2009  Florian Pose, Ingenieurgemeinschaft IgH
 *
 *  This file is part of the IgH EtherCAT Master.
 *
 *  The IgH EtherCAT Master is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License version 2, as
 *  published by the Free Software Foundation.
 *
 *  The IgH EtherCAT Master is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 *  Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with the IgH EtherCAT Master; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 *  ---
 *
 *  The license mentioned above concerns the source code only. Using the
 *  EtherCAT technology and brand is only permitted in compliance with the
 *  industrial property and similar rights of Beckhoff Automation GmbH.
 *
 ****************************************************************************/

#include <getopt.h>
#include <libgen.h> // basename()
#include <stdlib.h>

#include <iostream>
#include <iomanip>
using namespace std;

#include "CommandAlias.h"
#include "CommandConfig.h"
#include "CommandCStruct.h"
#include "CommandData.h"
#include "CommandDebug.h"
#include "CommandDiag.h"
#include "CommandDomains.h"
#include "CommandDownload.h"
#ifdef EC_EOE
#include "CommandEoe.h"
#include "CommandEoeAddIf.h"
#include "CommandEoeDelIf.h"
#endif
#include "CommandFoeRead.h"
#include "CommandFoeWrite.h"
#include "CommandGraph.h"
#include "CommandIp.h"
#include "CommandMaster.h"
#include "CommandPdos.h"
#include "CommandRegRead.h"
#include "CommandRegWrite.h"
#include "CommandRegReadWrite.h"
#include "CommandReboot.h"
#include "CommandRescan.h"
#include "CommandSdos.h"
#include "CommandSiiRead.h"
#include "CommandSiiWrite.h"
#include "CommandSlaves.h"
#include "CommandSoeRead.h"
#include "CommandSoeWrite.h"
#include "CommandStates.h"
#include "CommandUpload.h"
#include "CommandVersion.h"
#include "CommandXml.h"

#include "MasterDevice.h"

/*****************************************************************************/

typedef list<Command *> CommandList;
CommandList commandList;

string binaryBaseName;
string commandName;
Command::StringVector commandArgs;

// option variables
string masters = "-"; // all masters
string positions = "-"; // all positions
string aliases = "-"; // all aliases
string domains = "-"; // all domains
string dataTypeStr;
Command::Verbosity verbosity = Command::Normal;
bool force = false;
bool emergency = false;
bool helpRequested = false;
bool reset = false;
string outputFile;
string skin;

/*****************************************************************************/

string usage()
{
    stringstream str;
    CommandList::const_iterator ci;
    size_t maxWidth = 0;

    for (ci = commandList.begin(); ci != commandList.end(); ci++) {
        if ((*ci)->getName().length() > maxWidth) {
            maxWidth = (*ci)->getName().length();
        }
    }

    str << "Usage: " << binaryBaseName << " <COMMAND> [OPTIONS] [ARGUMENTS]"
        << endl << endl
        << "Commands (can be abbreviated):" << endl;

    str << left;
    for (ci = commandList.begin(); ci != commandList.end(); ci++) {
        str << "  " << setw(maxWidth) << (*ci)->getName()
            << "  " << (*ci)->getBriefDescription() << endl;
    }

    str << endl
        << "Global options:" << endl
        << "  --master  -m <master>  Comma separated list of masters" << endl
        << "                         to select, ranges are allowed." << endl
        << "                         Examples: '1,3', '5-7,9', '-3'." << endl
        << "                         Default: '-' (all)."
        << endl
        << "  --force   -f           Force a command." << endl
        << "  --quiet   -q           Output less information." << endl
        << "  --verbose -v           Output more information." << endl
        << "  --help    -h           Show this help." << endl
        << endl
        << Command::numericInfo()
        << endl
        << "Call '" << binaryBaseName
        << " <COMMAND> --help' for command-specific help." << endl
        << endl
        << "Send bug reports to " << PACKAGE_BUGREPORT << "." << endl;

    return str.str();
}

/*****************************************************************************/

void getOptions(int argc, char **argv)
{
    int c, argCount;
    stringstream str;

    static struct option longOptions[] = {
        //name,         has_arg,           flag, val
        {"master",      required_argument, NULL, 'm'},
        {"alias",       required_argument, NULL, 'a'},
        {"position",    required_argument, NULL, 'p'},
        {"domain",      required_argument, NULL, 'd'},
        {"type",        required_argument, NULL, 't'},
        {"output-file", required_argument, NULL, 'o'},
        {"skin",        required_argument, NULL, 's'},
        {"emergency",   no_argument,       NULL, 'e'},
        {"force",       no_argument,       NULL, 'f'},
        {"reset",       no_argument,       NULL, 'r'},
        {"quiet",       no_argument,       NULL, 'q'},
        {"verbose",     no_argument,       NULL, 'v'},
        {"help",        no_argument,       NULL, 'h'},
        {}
    };

    do {
        c = getopt_long(argc, argv, "m:a:p:d:t:o:s:efrqvh", longOptions, NULL);

        switch (c) {
            case 'm':
                masters = optarg;
                break;

            case 'a':
                aliases = optarg;
                break;

            case 'p':
                positions = optarg;
                break;

            case 'd':
                domains = optarg;
                break;

            case 't':
                dataTypeStr = optarg;
                break;

            case 'o':
                outputFile = optarg;
                break;

            case 's':
                skin = optarg;
                break;

            case 'e':
                emergency = true;
                break;

            case 'f':
                force = true;
                break;

            case 'r':
                reset = true;
                break;

            case 'q':
                verbosity = Command::Quiet;
                break;

            case 'v':
                verbosity = Command::Verbose;
                break;

            case 'h':
                helpRequested = true;
                break;

            case '?':
                cerr << endl << usage();
                exit(1);

            default:
                break;
        }
    }
    while (c != -1);

    argCount = argc - optind;

    if (!argCount) {
        if (helpRequested) {
            cout << usage();
            exit(0);
        } else {
            cerr << "Please specify a command!" << endl
                << endl << usage();
            exit(1);
        }
    }

    commandName = argv[optind];
    while (++optind < argc)
        commandArgs.push_back(string(argv[optind]));
}

/****************************************************************************/

list<Command *> getMatchingCommands(const string &cmdStr)
{
    CommandList::iterator ci;
    list<Command *> res;

    // see if there's an exact match
    for (ci = commandList.begin(); ci != commandList.end(); ci++) {
        if ((*ci)->matches(cmdStr)) {
            res.push_back(*ci);
            break;
        }
    }
    
    if (!res.size()) { // nothing found
        // find matching commands from beginning of the string
        for (ci = commandList.begin(); ci != commandList.end(); ci++) {
            if ((*ci)->matchesSubstr(cmdStr)) {
                res.push_back(*ci);
            }
        }
        
        if (!res.size()) { // nothing found
            // find /any/ matching commands
            for (ci = commandList.begin(); ci != commandList.end(); ci++) {
                if ((*ci)->matchesAbbrev(cmdStr)) {
                    res.push_back(*ci);
                }
            }
        }
    }

    return res;
}

/****************************************************************************/

int main(int argc, char **argv)
{
    int retval = 0;
    list<Command *> matchingCommands;
    list<Command *>::const_iterator ci;
    Command *cmd;

    binaryBaseName = basename(argv[0]);

    commandList.push_back(new CommandAlias());
    commandList.push_back(new CommandConfig());
    commandList.push_back(new CommandCStruct());
    commandList.push_back(new CommandData());
    commandList.push_back(new CommandDebug());
    commandList.push_back(new CommandDiag());
    commandList.push_back(new CommandDomains());
    commandList.push_back(new CommandDownload());
#ifdef EC_EOE
    commandList.push_back(new CommandEoe());
    commandList.push_back(new CommandEoeAddIf());
    commandList.push_back(new CommandEoeDelIf());
#endif
    commandList.push_back(new CommandFoeRead());
    commandList.push_back(new CommandFoeWrite());
    commandList.push_back(new CommandGraph());
    commandList.push_back(new CommandIp());
    commandList.push_back(new CommandMaster());
    commandList.push_back(new CommandPdos());
    commandList.push_back(new CommandRegRead());
    commandList.push_back(new CommandRegWrite());
    commandList.push_back(new CommandRegReadWrite());
    commandList.push_back(new CommandReboot());
    commandList.push_back(new CommandRescan());
    commandList.push_back(new CommandSdos());
    commandList.push_back(new CommandSiiRead());
    commandList.push_back(new CommandSiiWrite());
    commandList.push_back(new CommandSlaves());
    commandList.push_back(new CommandSoeRead());
    commandList.push_back(new CommandSoeWrite());
    commandList.push_back(new CommandStates());
    commandList.push_back(new CommandUpload());
    commandList.push_back(new CommandVersion());
    commandList.push_back(new CommandXml());

    getOptions(argc, argv);

    matchingCommands = getMatchingCommands(commandName);

    if (matchingCommands.size()) {
        if (matchingCommands.size() == 1) {
            cmd = matchingCommands.front();
            if (!helpRequested) {
                try {
                    cmd->setMasters(masters);
                    cmd->setVerbosity(verbosity);
                    cmd->setAliases(aliases);
                    cmd->setPositions(positions);
                    cmd->setDomains(domains);
                    cmd->setDataType(dataTypeStr);
                    cmd->setOutputFile(outputFile);
                    cmd->setSkin(skin);
                    cmd->setEmergency(emergency);
                    cmd->setForce(force);
                    cmd->setReset(reset);
                    cmd->execute(commandArgs);
                } catch (InvalidUsageException &e) {
                    cerr << e.what() << endl << endl;
                    cerr << cmd->helpString(binaryBaseName);
                    retval = 1;
                } catch (CommandException &e) {
                    cerr << e.what() << endl;
                    retval = 1;
                } catch (MasterDeviceException &e) {
                    cerr << e.what() << endl;
                    retval = 1;
                }
            } else {
                cout << cmd->helpString(binaryBaseName);
            }
        } else {
            cerr << "Ambiguous command abbreviation! Matching:" << endl;
            for (ci = matchingCommands.begin();
                    ci != matchingCommands.end();
                    ci++) {
                cerr << (*ci)->getName() << endl;
            }
            cerr << endl << usage();
            retval = 1;
        }
    } else {
        cerr << "Unknown command " << commandName << "!" << endl
            << endl << usage();
        retval = 1;
    }

    return retval;
}

/****************************************************************************/
